python/tank/util/yaml_cache.py

```diff
@@ -18,17 +18,124 @@
+        """
+        self._path = path
+        self._data = data
+        try:
+            self._stat = stat or os.stat(path)
+        except Exception, exc:
+            raise TankUnreadableFileError("Unable to stat file '%s': %s" % (path, exc))
+
+    @property
+    def data(self):
+        """The item's data."""
+        return self._data
+
+    @data.setter
+    def data(self, config_data):
+        self._data = config_data
+
+    @property
+    def path(self):
+        """The path to the file on disk that the item was sourced from."""
+        return self._path
+
+    @property
+    def stat(self):
+        """The stat of the file on disk that the item was sourced from."""
+        return self._stat
+
+    def given_item_newer(self, other):
+        """
+        Tests whether the given item is newer than this.
+
+        :param other:   The CacheItem to test against.
+        :raises:        TypeError: Given item is not a CacheItem.
+        :returns:       bool, True if other is newer, False if not.
+        """
+        if not isinstance(other, CacheItem):
+            raise TypeError("Given item must be of type CacheItem.")
+        return other.stat.st_mtime > self.stat.st_mtime
+
+    def size_differs(self, other):
+        """
+        Tests whether the file size of the given item differs from this item.
+
+        :param other:   The CacheItem to test against.
+        :raises:        TypeError: Given item is not a CacheItem.
+        :returns:       bool, True if other is a different size on disk, False if not.
+        """
+        if not isinstance(other, CacheItem):
+            raise TypeError("Given item must be of type CacheItem.")
+        return other.stat.st_size != self.stat.st_size
+
+    def __eq__(self, other):
+        if not isinstance(other, CacheItem):
+            raise TypeError("Given item must be of type CacheItem.")
+        return (other.stat.st_mtime == self.stat.st_mtime and not self.size_differs(other))
+
+    def __getitem__(self, key):
+        # Backwards compatibility just in case something outside
+        # of this module is expecting the old dict structure.
+        if key == "modified_at":
+            return self.stat.st_mtime
+        elif key == "file_size":
+            return self.stat.st_size
+        elif key == "data":
+            return self._data
+        else:
+            return getattr(self._data, key)
+
+    def __str__(self):
+        return str(self.path)
 
 class YamlCache(object):
     """
     Main yaml cache class
     """
-    def __init__(self):
+    # The cache will be a singleton. At the bottom of this
+    # module is defined a global YamlCache object, which
+    # will then be shared if/when other instantiations occur.
+    # It's likely that most all code will be referencing the
+    # global variable for the cache, but this will ensure that
+    # even if that's not the case we'll still only have a single
+    # cache used throughout.
+    _instance = None
```

Comment:
Probably so. I added that early on not trusting that everything in core made use of the globally-defined cache. Subsequent digging around in various parts of the code have made me less concerned about that. I've not since found any outliers instantiating their own cache directly.